crates/core/tedge_api/src/entity_store.rs

```diff
@@ -424,47 +440,75 @@ impl EntityRegistrationMessage {
     /// MQTT message is an entity registration message if
     /// - published on a prefix of `te/+/+/+/+`
     /// - its payload contains a registration message.
+    // TODO: replace option with proper error handling
+    // TODO: this is basically manual Deserialize implementation, better impl
+    // Serialize/Deserialize.
     #[must_use]
     pub fn new(message: &Message) -> Option<Self> {
         let payload = parse_entity_register_payload(message.payload_bytes())?;
 
-        let r#type = payload
-            .get("@type")
-            .and_then(|t| t.as_str())
-            .map(|t| t.to_owned())?;
+        let JsonValue::Object(mut properties) = payload else {
+            return None;
+        };
+
+        let Some(JsonValue::String(r#type)) = properties.remove("@type") else {
+            return None;
+        };
+
         let r#type = match r#type.as_str() {
             "device" => EntityType::MainDevice,
             "child-device" => EntityType::ChildDevice,
             "service" => EntityType::Service,
             _ => return None,
         };
 
+        let parent = properties.remove("@parent");
+        let parent = if let Some(parent) = parent {
+            let JsonValue::String(parent) = parent else {
+                return None;
+            };
+            let Ok(parent) = parent.parse() else {
+                return None;
+            };
+            Some(parent)
+        } else {
+            None
+        };
+
         let parent = if r#type == EntityType::ChildDevice || r#type == EntityType::Service {
-            payload
-                .get("@parent")
-                .and_then(|p| p.as_str())
-                .and_then(|p| p.parse().ok())
+            parent
         } else {
             None
         };
 
-        let entity_id = payload
-            .get("@id")
-            .and_then(|id| id.as_str())
-            .map(|id| id.into());
+        let entity_id = properties.remove("@id");
+        let entity_id = if let Some(entity_id) = entity_id {
+            let JsonValue::String(entity_id) = entity_id else {
+                return None;
+            };
+            Some(entity_id.into())
+        } else {
+            None
+        };
 
         let topic_id = message
             .topic
             .name
             .strip_prefix(MQTT_ROOT)
             .and_then(|s| s.strip_prefix('/'))?;
 
+        let other = JsonValue::Object(properties);
+
+        assert_eq!(other.get("@id"), None);
+        assert_eq!(other.get("@type"), None);
+        assert_eq!(other.get("@parent"), None);
```

Comment:
These assertions should never fail, as I removed these entries before in a `properties` map, but I'd like to keep them as a final defensive check that the returned object doesn't have those properties. It's also more clear when reading that "`other` should not have these properties" than when it's scattered above in `.remove()` calls.

[Asserts are not only useful for tests](https://tratt.net/laurie/blog/2023/rusts_two_kinds_of_assert_make_for_better_code.html). Here I used them more like a comment that "these things should be true" that also executes and checks that these things are indeed true. Perhaps here it's a little trivial, because I do remove these entries above, so I could've used `debug_assert`, but it's not like this is a hot loop or anything so these asserts should be essentially free.

Reply:
Thanks for that article. I didn't know about `debug_assert` before, and yes I agree that entity registrations are not that frequent to warrant any performance worries.